### PR TITLE
Add --home to specify location for persistent data

### PR DESCRIPTION
**Pull request by tangledhelix**
## Changes

Adds a `--home` flag, taking a directory as an argument, which specifies a directory for storing persistent user data. At this time, three things would be stored in this directory:

 - `setting.rc`
 - `header.txt`
 - `labels_{lang}.rc`

In the case of a corrupted `setting.rc`, then this location is also where `setting.err` might be written.

This flag is optional, the default behavior is preserved.

Changes argument parsing to use Getopt::Long. Options `--home` and `--runtests` are parsed with Getopt::Long. If a filename is specified, that is still handled by reading `@ARGV`.

**Note that this means a new arg `--runtests` replaces the older form `runtests`.**

Also adds helper functions to return various paths:
 - `header.txt`
 - `headerdefault.txt`
 - `labels_{lang}.default.rc`
 - `labels_{lang}.rc`
 - `setting.rc`

Some of the helpers are to simplify things with the variable location (whether `--home` is in use or not), some are just for the sake of convenience.

## Purposes

As described in #954, one purpose is to allow an easy way for users to carry their settings from one release to the next without manually copying files around.

I am working on a macOS app bundle to simplify the install and upgrade process for Mac users. Since Mac apps must be signed, their contents can't change; the `--home` flag permits user data to live outside of the application's directory, making this possible.

## Testing

I tested these changes on macOS. Since macOS uses a similar filesystem interface, Linux probably also works, but I didn't test directly.

I have no Windows machines handy to test with, so I'd appreciate if someone could give it a try on Windows.

### CODE REVIEW

**Comment by tangledhelix** [comment]
I fixed what I could for perltidy, but now it's hung up on what appears to be whitespace? I found one line with trailing whitespace (fixed), but now it's complaining about two locations that appear to already be correct... if someone else can spot what it's unhappy about, please let me know...

**Comment by tangledhelix** [comment]
@windymilla The Actions workflow is only actually complaining about Utilities.pm, and only line 354... but there **is** a blank line after `path_labels`, and there's no trailing space on it, etc... for the life of me I can't see what it's complaining about here...

https://github.com/DistributedProofreaders/guiguts/runs/7398845301?check_suite_focus=true

**Comment by tangledhelix** [comment]
I decided to take @windymilla 's suggestion and let format_files do what it wanted... but now there are many more errors, so that seems to have gone worse than where I started from 😣

**Comment by windymilla** [comment]
OK, partial memory pointed me in the right direction and found the explanation and solution

The [README](https://github.com/DistributedProofreaders/guiguts/blob/master/README.md) says for developers,

> Guiguts uses [perltidy](https://metacpan.org/pod/Perl::Tidy) for consistent styling using configuration options in `.perltidyrc`. 
> You can install perltidy with `cpanm --installdeps --notest --with-develop .` including the dot at the end.

On running that, it installed perltidy version 20210111, which is the one used by the Actions, as can be seen in the log for the `perltidy --version` Action. This local command and the Action are both controlled by [cpanfile](https://github.com/DistributedProofreaders/guiguts/blob/master/cpanfile) which contains the perltidy version number.

Using that version of perltidy, all of your files check OK, except Utilities, which requires a blank line after line 354
`sub path_labels {`
and some spaces removing from the otherwise blank line 909 (which becomes 910) with the above edit), just before the `::composeinitialize();` line.

Hope that helps - if you install the 20210111 version of perltidy, you should be able to use `format_files` script to do the tidying for you in future. At some point we can pin the version of perltidy to a later release for Actions and all GG developers would need to re-run the above command to install the correct one locally.

**Comment by tangledhelix** [comment]
Thanks @windymilla - worked like a charm. The CI is all green now.

**Comment by srjfoo** [comment]
First round of testing (sorry to be so slow).

Downloaded code to the desktop. This time I just used the default `perl guiguts.pl &` command to open it, to see what the default does. I opened GG, typed in some text, saved, and although the text file saved, GG quit:
```
[1]+  Segmentation fault: 11  perl guiguts.pl
```
I brought GG back up, opened the file, typed in some more text, and saved successfully; no segfault. However, when I quit GG, brought it up again, and saved another new document, I got a segfault again. `setting.rc` and `header.txt` were created successfully, but `labels_en.rc` was not.

More testing mañana.

**Comment by tangledhelix** [comment]
@srjfoo I can reproduce what you saw, but I have two things to add:

1. For me, `data/labels_en.rc` **was** created
2. I get that behavior also on the master branch, so whatever's going on there isn't related to this PR as far as I can tell

**Comment by srjfoo** [comment]
> 

Perhaps I got the wrong file, then. The zip I downloaded is `guiguts-home_flag.zip`, and I got it by going to [tangledhelix:home_flag](https://github.com/tangledhelix/guiguts/tree/home_flag), clicking on the "Code" button and selecting "Download ZIP" (rather than doing this through my main dev install). I then `cd` into the `src` subdirectory and run guiguts from there. So, perhaps I need a change in procedure -- or I'm downloading the wrong thing?

**Comment by srjfoo** [comment]
> 

Okay -- a little bit more awake now. 😁 

I have now tested with `master`, also, and confirm that I get a segfault on initial save using the default "save" on `master`, also. If, instead of a "save", I do a `save as` for that initial save, it does not segfault.

I've also confirmed that `labels_en.rc` exists -- it's just not where I expected it to be. It seems to be staying in the `data` subfolder, and overwriting the `labels_en_default.rc` file, instead of landing in the directory I invoke gg from, whereas in my previous test, when I specified a directory for the settings files, it was saved in that directory.

I just checked my master dev branch (which has _not_ been updated to the current `master`) and it also saves `labels_en.rc` in the `data` subfolder, but does not appear to overwrite the default? At least, I assume so, but I may be mistaken about that. At any rate, both `labels_en.rc` and `labels_en_default.rc` are there.

One final note: I brought up gg using my dev master and tried the save button to do the initial save. The file was created, but the program crashed, but this time, the error message was:
```
[1]+  Bus error: 10           perl guiguts.pl
```

**Comment by rtonsing** [comment]
Did some testing in Windows 11, seems to work OK.

**Comment by windymilla** [comment]
> One final note: I brought up gg using my dev master and tried the save button to do the initial save. The file was created, but the program crashed, but this time, the error message was:
> 
> ```
> [1]+  Bus error: 10           perl guiguts.pl
> ```
Also reproduced on Windows - have added issue #960 since it doesn't appear related to this PR

**Comment by windymilla** [comment]
Initial tests on Windows 11 working for me - needs a bit more time

**Comment by tangledhelix** [comment]
> I've also confirmed that `labels_en.rc` exists -- it's just not where I expected it to be. It seems to be staying in the `data` subfolder, and overwriting the `labels_en_default.rc` file, instead of landing in the directory I invoke gg from, whereas in my previous test, when I specified a directory for the settings files, it was saved in that directory.

The path difference is known / intended. When `--home` is used, I wrote all the files in the home directory without any subdirectories. But if `--home` is not used, I left the paths the way they have always been. In general, my design principle here was to not change anything unless `--home` was specified.

The only exception to that rule was that `runtests` became `--runtests` when Getopt::Long was introduced. But that seemed acceptable (perhaps @windymilla or @cpeel can tell me if it's not).

If `labels_en_default.rc` is in fact being overwritten, that is not intended. The desired behavior is:

```
if labels_{booklang}.rc does not exist:
    if labels_{booklang}_default.rc exists:
        copy labels_{booklang}_default.rc to labels_{booklang}.rc
    else
        # if no defaults for this language, then use English as global default
        copy labels_en_default.rc to labels_{booklang}.rc
    end
end
```

**Comment by srjfoo** [comment]
> If `labels_en_default.rc` is in fact being overwritten, that is not intended. The desired behavior is:
> 
> ```
> if labels_{booklang}.rc does not exist:
>     if labels_{booklang}_default.rc exists:
>         copy labels_{booklang}_default.rc to labels_{booklang}.rc
>     else
>         # if no defaults for this language, then use English as global default
>         copy labels_en_default.rc to labels_{booklang}.rc
>     end
> end
> ```

My bad on that -- I must have been blind. I ran the tests again (deliberately invoking the bus error/segfault) and `labels_en_default.rc` is not actually overwritten.

Thanks for the explanation of where things live -- that makes sense.

File: src/lib/Guiguts/Utilities.pm [hunk 1]
```diff
@@ -337,6 +338,39 @@ sub dos_path {
...
+
+sub path_defaulthtmlheader {
+    return 'headerdefault.txt';
+}
+
+sub path_labels {
+
+    # If we're using --home (a separate data directory), then we store the
+    # labels file directly there
+    if ( $::lglobal{homedirectory} ne $::lglobal{guigutsdirectory} ) {
```
Comment (windymilla) [inline src/lib/Guiguts/Utilities.pm:358]: Although it would be a bizarre thing to do with our current plans for the use of this feature, I could envisage a future use might end up using the `--home` argument, but with the `guigutsdirectory` directory as the argument. On Windows, folders resolve OK even if the case is wrong, so passing in `\dp\guiguts` or \DP\guiguts` as the `homedirectory` refer to the same folder. Only one of these will be `eq` to the  `guigutsdirectory`, so the treatment of the labels file will be different in the two cases.
I'm not sure this matters, and may not be worth the extra code complexity of saying "On Windows, do a case-insensitive comparison". I did look at `URI::File` but although it gives a "canonical" version of a folder name, it doesn't change case. 
So, just an observation, unless you have a good idea for avoiding it simply.

Comment (tangledhelix) [reply src/lib/Guiguts/Utilities.pm:358]: I see what you mean. Macs also have case-insensitive filesystems(*), so on both Mac and Windows we should compare the path without considering case (i.e. with `lc`). I'll look at it this weekend if I have time to test it.

(*) There's technically an edge case where someone could run Guiguts from a Mac case-sensitive volume, but you'd have to really go out of your way to do this and I don't think it's worth checking for it.

Comment (srjfoo) [reply src/lib/Guiguts/Utilities.pm:358]: > (*) There's technically an edge case where someone could run Guiguts from a Mac case-sensitive volume, but you'd have to really go out of your way to do this and I don't think it's worth checking for it.

I agree, but it might be worth a comment line that a case-insensitive OS is assumed? Periodically, I've been tempted to do a case-sensitive install, but have always chickened out. 😁

Comment (tangledhelix) [reply src/lib/Guiguts/Utilities.pm]: I just pushed a commit that does the case-insensitive comparison when one of (OS_WIN, OS_MAC) are true.

Comment (windymilla) [reply src/lib/Guiguts/Utilities.pm]: That all looks good to me, thanks

**Review by windymilla — APPROVED** [review]
I'm happy with all of this now if you are too @srjfoo

**Comment by srjfoo** [comment]
> 

Looks good to me. I went back to my original quick test of changing background color and button highlight color and show/hide line numbers, and all that seemed to be working just fine (I haven't tested out everything; just those three).

I created two different "`home`" directories, each with different settings, one all-caps, and flipped between them, invoking the all-capped directory with lower-case. Settings on bringing up GG were correct for the setting file I used.

If there's anything else I should be testing, let me know.

**Review by srjfoo — APPROVED** [review]
